### PR TITLE
Feature/registry upgrade

### PR DESCRIPTION
**Pull request by yippee-ki-yay**
Add ability for users to change DSProxy owner in one tx and also keep a pool of premade dsproxies for easier on boarding.
The new DFSProxyRegistryController will be the owner of the DFSRegistry for this to work.

### CODE REVIEW

File: contracts/actions/utils/ChangeProxyOwner.sol [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    //////////////////////////// ACTION LOGIC ////////////////////////////
+
+    function _changeOwner(address _newOwner) internal {
+        require(_newOwner != address(0), "Owner is empty address");
+
+        DSAuth(address(this)).setOwner(_newOwner);
+        
+        dfsRegController.changeOwnerInDFSRegistry(_newOwner);
+    }
```
Comment (majkic99) [inline contracts/actions/utils/ChangeProxyOwner.sol:50]: add Logger in action

File: contracts/utils/DFSProxyRegistryController.sol [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    event NewProxy(address, address);
+    event ChangedOwner(address, address);
+
+    /// @notice User calls from EOA to build a new DFS registred proxy
+    function addNewProxy() public {
+        address newProxy = getFromPoolOrBuild(msg.sender);
+        DFSProxyRegistry(DFS_PROXY_REGISTRY_ADDR).addAdditionalProxy(msg.sender, address(newProxy));
+
+        emit NewProxy(msg.sender, newProxy);
```
Comment (majkic99) [inline contracts/utils/DFSProxyRegistryController.sol:25]: maybe return newProxy address?

File: contracts/utils/DFSProxyRegistryController.sol [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    /// @dev List of prebuild proxies the users can claim to save gas
+    address[] public proxyPool;
+
+    event NewProxy(address, address);
+    event ChangedOwner(address, address);
+
+    /// @notice User calls from EOA to build a new DFS registred proxy
+    function addNewProxy() public {
+        address newProxy = getFromPoolOrBuild(msg.sender);
+        DFSProxyRegistry(DFS_PROXY_REGISTRY_ADDR).addAdditionalProxy(msg.sender, address(newProxy));
```
Comment (majkic99) [inline contracts/utils/DFSProxyRegistryController.sol:23]: casting address -> address

File: contracts/actions/utils/ChangeProxyOwner.sol [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        address newOwner = parseInputs(_callData);
+
+        newOwner = _parseParamAddr(newOwner, _paramMapping[0], _subData, _returnValues);
+
+        _changeOwner(newOwner);
+
+        return bytes32(bytes20(newOwner));
+    }
+
+    // solhint-disable-next-line no-empty-blocks
```
Comment (majkic99) [inline contracts/actions/utils/ChangeProxyOwner.sol:30]: // solhint-disable not needed

File: contracts/utils/DFSProxyRegistryController.sol [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        if (additionalProxies.length == 0) {
+            return proxies;
+        }
+
+        for (uint256 i = 0; i < additionalProxies.length; ++i) {
+            proxies[i + 1] = additionalProxies[i];
+        }
+
+        return proxies;
+    }
```
Comment (majkic99) [inline contracts/utils/DFSProxyRegistryController.sol:81]: add view function that returns proxyPool.length

**Review by majkic99 — CHANGES_REQUESTED** [review]
looks safe, some small changes needed

**Review by majkic99 — APPROVED** [review]

**Review by nklipa13 — APPROVED** [review]
lgtm

**Pull request merged**